packages/meditrak-server/src/routes/exportSurveys/assertCanExportSurveys.js

```diff
@@ -0,0 +1,52 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2020 Beyond Essential Systems Pty Ltd
+ */
+
+import { flattenDeep } from 'lodash';
+
+import { TUPAIA_ADMIN_PANEL_PERMISSION_GROUP } from '../../permissions';
+
+const getCountryNames = async (models, surveyCountryCodes) => {
+  if (surveyCountryCodes) {
+    const surveyCountries = await models.country.find({ code: surveyCountryCodes });
+    return surveyCountries.map(c => c.name).join(',');
+  }
+
+  return 'any countries';
+};
+
+export const assertCanExportSurveys = async (accessPolicy, models, surveys) => {
+  const surveyCountryIds = flattenDeep(surveys.map(s => s.country_ids));
+  const surveyPermissionGroupIds = flattenDeep(surveys.map(s => s.permission_group_id));
+  const countryCodeById = await models.country.getCountryCodeById(surveyCountryIds);
+  const permissionGroupNameById = await models.permissionGroup.getPermissionGroupNameById(
+    surveyPermissionGroupIds,
+  );
+
+  for (const survey of surveys) {
+    const surveyPermissionGroupName = permissionGroupNameById[survey.permission_group_id];
+    const surveyCountryCodes =
+      survey.country_ids && survey.country_ids.length
+        ? survey.country_ids.map(id => countryCodeById[id])
+        : null;
+
+    if (!accessPolicy.allowsSome(surveyCountryCodes, surveyPermissionGroupName)) {
+      const surveyCountryNamesString = await getCountryNames(models, surveyCountryCodes);
+
+      throw new Error(
+        `Need ${surveyPermissionGroupName} access to ${surveyCountryNamesString} to export the survey ${survey.name}`,
```

Comment:
I do find this error message a little confusing tbh. So if I am correct the problem is that:
- There is an export survey, who's permission group has no access to any of the countries that it is exporting for?

This seems like a configuration issue? If so, I think perhaps an error message that instructs the user to contact their administrator might be a bit nicer, eg:
`The survey '${survey.name}' does not have permission to export data for: ${surveyCountryNamesString}. Please contact the project coordinators to resolve this issue.`

Sorry if I'm incorrect about this being a configuration issue, I'm just a little confused how a user should ever end up in this scenario

Reply:
Sorry for my late reply!!! Thank you for your suggestion!

To explain this, each survey has a permission group and some countries attached to it. What I want to check here is if users have that survey's permission group access to at least 1 of the survey's attached countries:

Eg: Survey BCD has:
- permissionGroup: Donor
- countries: [Tonga, Demo Land]

We want to check if the user has Donor access to either Tonga or Demo Land. If not, the user cannot export the survey

So I think this is not quite a configuration issue? And the message `The survey '${survey.name}' does not have permission to export data for: ${surveyCountryNamesString}` is not 100% accurate. To fix this issue, the user will have to request/add the Donor permission to either Tonga or Demo Land

Reply:
Ahh okay I see your point 😄 So would an accurate message be something like:
`The survey's ('${survey.name}') permission group ('${surveyPermissionGroupName}') does not have permission to export data for: ${surveyCountryNamesString}. Please contact the project coordinators to resolve this issue.`?

I guess the reason I'm being so nitpicky about this message is that with the current wording I think the user may feel that that it's an issue with their own permissions. For instance if I saw: `Need Donor access to Vanuatu to export the survey BCD` then I may think that my user needs `Donor` access or something. But in reality, it's an issue with how the survey/survey permission group is configured, and has nothing to do with the user's permissions.

Do you agree with my assessment? I'm not too fussed about the exact wording of the error message, as long as it indicates to the user that it's the survey that needs its permissions fixed, and that they should contact their admin to do this.

Reply:
I think I see your point now! 

> For instance if I saw: Need Donor access to Vanuatu to export the survey BCD then I may think that my user needs Donor access or something. But in reality, it's an issue with how the survey/survey permission group is configured, and has nothing to do with the user's permissions.

You're right that when I was working on this I actually thought that if users cannot export a survey, they should request/add permission needed for this survey (instead of changing the survey config like you said). Hence that's why the error message was like that. If in real life scenarios, users tend to have the survey reconfigured instead of adding/request the permission they need (is this the case?), then I agree with you!

However, there's one difficulty with your suggested message that I can think of. 
> The survey's ('${survey.name}') permission group ('${surveyPermissionGroupName}') does not have permission to export data for: ${surveyCountryNamesString}. Please contact the project coordinators to resolve this issue?

In your suggested message above, `surveyCountryNamesString` is not the corresponding countries of the survey anymore (right now it is that case). It will have to be the countries that the users have the survey's permission access to, which we have to go through all the permissions that the user has and find the corresponding countries.

Eg: 
Users have Donor permission access to KI, LA, SB
Survey BCD has:
- permissionGroup: Donor
- countries: [Tonga, Demo Land]
Then we need to go through the access policy of the users, find any Donor permission they have and grab the corresponding countries

To make this easier, can we change this to 
`The survey '${survey.name}' has permission group ('${surveyPermissionGroupName}') access to only ${surveyCountryNamesString}. Please contact the project coordinators to resolve this issue`?

Reply:
Omg 🤦 sorry @chris-bes, I think I misunderstood the permissions model here.

>You're right that when I was working on this I actually thought that if users cannot export a survey, they should request/add permission needed for this survey (instead of changing the survey config like you said). Hence that's why the error message was like that.

I think you're correct with your initial thoughts. The user needs to extend their permissions to include the country and permission group. Given what I understand now, I'm happy with the message that exists currently, though if after all this conversation you wanted to re-word it at all I'm happy for that too 😄 

Sorry for creating all this fuss over nothing, thanks for being so patient with me ❤️